osbrain/tests/test_agent_async_requests.py

```diff
@@ -22,114 +23,113 @@ def on_error(agent):
 class Server(Agent):
     def on_init(self):
         self.received = []
+        self.blocked = False
 
 
 class Client(Agent):
     def on_init(self):
         self.received = []
 
 
-@pytest.fixture(scope='function')
-def server_client_late_reply_return():
-    def late_reply(agent, request):
-        agent.received.append(request)
-        time.sleep(1)
-        return 'x' + request
-
+def async_client_server(server_handler=append_received,
+                        client_handler=append_received):
     server = run_agent('server', base=Server)
     client = run_agent('client', base=Client)
 
-    addr = server.bind('ASYNC_REP', alias='replier', handler=late_reply)
-    client.connect(addr, alias='async', handler=append_received)
+    addr = server.bind('ASYNC_REP', alias='replier', handler=server_handler)
+    client.connect(addr, alias='async', handler=client_handler)
 
     return (server, client)
 
 
+def async_client_servers(server_handler=append_received,
+                         client_handler=append_received):
+    server1 = run_agent('server1', base=Server)
+    server2 = run_agent('server2', base=Server)
+    client = run_agent('client', base=Client)
+
+    addr1 = server1.bind('ASYNC_REP', alias='replier1', handler=server_handler)
+    addr2 = server2.bind('ASYNC_REP', alias='replier2', handler=server_handler)
+    client.connect(addr1, alias='async1', handler=client_handler)
+    client.connect(addr2, alias='async2', handler=client_handler)
+
+    return (server1, server2, client)
+
+
 @pytest.fixture(scope='function')
 def server_client_late_reply_delay():
     def late_reply(agent, delay):
         agent.received.append(delay)
         time.sleep(delay)
         return 'x' + str(delay)
 
-    server = run_agent('server', base=Server)
-    client = run_agent('client', base=Client)
+    return async_client_server(late_reply)
 
-    addr = server.bind('ASYNC_REP', alias='replier', handler=late_reply)
-    client.connect(addr, alias='async', handler=append_received)
 
-    return (server, client)
+def blocked_reply(agent, request):
+    agent.received.append(request)
+    agent.blocked = True
```

Comment:
Even if very unlikely, I would put `agent.blocked = True` before the append to avoid failing due to race conditions.

In example: in tests you set `blocked=False` just after checking the `received` attribute.